src/fsharp/MethodOverrides.fs

```diff
@@ -460,29 +460,34 @@ module DispatchSlotChecking =
-                                                // TODO: Move this to SR
-                                                errorR(Error((5000, sprintf "Interface member %A does not have a most specific implementation." (NicePrint.stringOfMethInfo amap reqdTyRange denv reqdSlot)), reqdTyRange))
-
-                                 if hasMostSpecificImplementation then                                 
-                                     match sortedSlots with
-                                     | [] ->
-                                         yield RequiredSlot(reqdSlot, not reqdSlot.IsAbstract)
-                                     | (_, bestSlot) :: _ ->                                      
-                                         yield bestSlot
-                                 else
-                                     // We make this optional because we already error'ed due to no most specific override.
-                                     // Try not to show more errors.
-                                     yield RequiredSlot(reqdSlot, true)
+                                
+                                 match sortedSlots with
+                                 | [] ->
+                                    yield RequiredSlot(reqdSlot, not reqdSlot.IsAbstract)
+                                 | (ty1, (RequiredSlot (_, isOptional1) as bestSlot)) :: sortedSlotsTail ->
+
+                                    // TODO: Add check for most specific implementation when an explicit implementation is given.
+                                    let mutable hasMostSpecificImplementation = true
+                                    let mutable isOptional1 = isOptional1
+                                    for (ty2, RequiredSlot (_, isOptional2)) in sortedSlotsTail do
+
+                                        let isNotPartOfHierarchy =
+                                            not (TypeFeasiblySubsumesType 0 g amap reqdTyRange ty1 CanCoerce ty2) && 
+                                            not (TypeFeasiblySubsumesType 0 g amap reqdTyRange ty2 CanCoerce ty1)
+
+                                        if isOptional1 && isOptional2 && hasMostSpecificImplementation && isNotPartOfHierarchy then
+                                            hasMostSpecificImplementation <- false
+
+                                        if isNotPartOfHierarchy then
+                                            isOptional1 <- if isOptional1 then true else isOptional2 
+
+                                    if hasMostSpecificImplementation then
+                                        yield bestSlot
+                                    else
+                                        // TODO: Move this to SR
+                                        errorR(Error((5000, sprintf "Interface member %A does not have a most specific implementation." (NicePrint.stringOfMethInfo amap reqdTyRange denv reqdSlot)), reqdTyRange))
```

Comment:
I assume we'll have a different value than `5000`?

Reply:
This section is still WIP. I appreciate the review, but this section could all change once I finalize the rules and do cleanup and optimization. Yes, the value will be different and I just have it as a placeholder for now.